### DISCUSSION TITLE
How to avoid "Error: Hydration failed because the initial UI does not match what was rendered on the server." in Next.js

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#38002] [Help]
I am getting this expected error:

```
Unhandled Runtime Error
Error: Hydration failed because the initial UI does not match what was rendered on the server.
```

And

```
Unhandled Runtime Error
Error: Text content does not match server-rendered HTML.
```

This has happened to me twice so far in situations I would like to do it _anyways_, and ignore the error. How can I ignore the error, or in what ways can I accomplish the following.

In the first situation, I add a class name if it's a mobile view vs. a landscape/desktop view, which doesn't exist on the server. This throws the "Hydration failed" error because there is an additional class name added to the DOM on client-side rendering only. I do it in a `useLayoutEffect` because otherwise I get a jolt of desktop-styled content on mobile before the class name is added, if I just use `useEffect`.

In the second situation, I am doing an onboarding flow where you can't skip steps but can only go to the furthest step you have completed (and you can go back to previous steps). But there are no URL changes, it is all kept track of in `localStorage`, because what is linking to the middle of an onboarding flow going to really accomplish in the grand scheme of things? But at the same time I would like to catch common mistakes like accidentally refreshing the page, it should jump you back to where you were in the flow. Etc..

In that situation, say there are 4 onboarding steps, and I left off at the 3rd one. Well, on the server it is going to render the first step, but on the client it is going render the 3rd step, as expected. But it also throws that "Hydration failed" error. This is desired behavior though (at least I think?!). This is how I've always implemented it in the past outside of Next.js.

How can I either ignore this error, or workaround it?

I really like the idea of having server-side rendering, and for all intents and purposes, the server-side rendering is correct (it is stateless). But when it reaches the client, which has some state, it might slightly adjust the content of the page. I don't see why it needs to error, or how to avoid this problem. Any help would be appreciated! Thank you.

```

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64
    Binaries:
      Node: 17.7.2
      npm: 7.19.1
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.1.6
      react: 18.2.0
      react-dom: 18.2.0

```

**Comment by GabenGar** [comment]
In the first situation write the styles so they would depend on viewport instead of classname.
The second situation can be avoided by blocking the render until client render kicks in from `useEffect()`.

**Reply by icyJoseph** [reply]
Fully agreed on the first case. Though, with middleware you could rewrite to different versions of the page depending on the user agent, I'd still prefer CSS based solutions anytime.

For part two, I think another solution would be to use cookies. A simple http only cookie indicating the step is a proper replacement for local-storage driven CSR persistence. Just use a cookie, with GSSP, tells Next.js which step to render/send JSON for to the client.

Another solution for part two would be to pre-render the shell of the onboarding and let the client figure out which step to render, this would require a `show` flag to be set inside a useEffect, as suggested by GabenGar.
